### PR TITLE
tectonic 0.1.6 (new formula)

### PR DESCRIPTION
**Pull request by iKevinY**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

Thanks to #14490 getting merged in, we're closer to a functional formula for [Tectonic](https://github.com/tectonic-typesetting/tectonic)! I'm submitting this as a WIP to move Homebrew-specific discussion out of tectonic-typesetting/tectonic#4, but there's a couple of things that need still need to be addressed:

- [ ] Ideally, the `test` block should invoke `tectonic`. Unfortunately, a bunch of rather large TeX packages need to be downloaded in order to process even a minimal (empty) document.
- [x] `brew audit` fails because the `harfbuzz` dependency needs to be installed `with-graphite2`.

### CODE REVIEW

**Comment by ilovezfs** [comment]
@iKevinY this will take care of harfbuzz: https://github.com/Homebrew/homebrew-core/pull/14583

**Comment by ilovezfs** [comment]
@iKevinY regarding the audit

>Dependency 'icu4c' may be unnecessary as it is provided by macOS; try to build this formula without it.

Have you tested a build that doesn't depend on the `icu4c` formula?

**Comment by ilovezfs** [comment]
@iKevinY the build fails on Yosemite, and also fails on El Capitan unless I have Xcode not just the CLT.

**Comment by ilovezfs** [comment]
@iKevinY I tried it with paper.tex from the repo, and it seems to error out: https://gist.github.com/ilovezfs/befd84e087fad9c72422e90b09418b70

```
iMac-TMP:footec joe$ tectonic paper.tex 
note: this is a BETA release; report issues at https://github.com/tectonic-typesetting/tectonic/issues
Running TeX ...
error: something bad happened inside TeX; its output follows:

===============================================================================
(paper.tex
LaTeX2e <2016/03/31>
Babel <3.9r> and hyphenation patterns for 83 language(s) loaded.
(aastex6.cls

LaTeX Warning: You have requested document class `aastex6',
               but the document class provides `aastex'.

Document Class: aastex 2016/02/16 1.0/AAS markup document class
(times.sty) (lineno.sty) (revtex4-1.cls
Document Class: revtex4-1 2010/07/25/20:33:00 4.1r (http://publish.aps.org/revt
ex4/ for documentation)
ltxutil[2010/07/25/20:33:00 4.1r utilities package (portions licensed from W. E
. Baxter web at superscript.com)]
ltxfront[2010/07/25/20:33:00 4.1r frontmatter package (AO,DPC)]
ltxgrid[2010/07/25/20:33:00 4.1r page grid package (portions licensed from W. E
. Baxter web at superscript.com)]

Class revtex4-1 Warning: Failed to recognize \@vspace, \@vspacer, \@no@pgbk, \@
newline, and \\; no patches applied. Please get a more up-to-date class, .

(aps4-1.rtx) (aps10pt4-1.rtx) (textcase.sty) (url.sty) (natbib.sty)
(revsymb4-1.sty)) (latexsym.sty) (graphicx.sty (keyval.sty) (graphics.sty
(trig.sty) (graphics.cfg) (xetex.def (infwarerr.sty) (ltxcmds.sty))))
(amssymb.sty (amsfonts.sty)) (epsf.sty
This is `epsf.tex' v2.7.4 <14 February 2011>
) (longtable.sty) (hyperref.sty (hobsub-hyperref.sty (hobsub-generic.sty))
(ifxetex.sty) (auxhook.sty) (kvoptions.sty) (pd1enc.def) (hyperref.cfg))

Package hyperref Message: Driver (autodetected): hxetex.

(hxetex.def (puenc.def) (stringenc.sty) (rerunfilecheck.sty)) (xcolor.sty
(color.cfg)) (array.sty) (ulem.sty))===============================================================================

error: failed to open input file "mode"
```

**Comment by iKevinY** [comment]
> this will take care of harfbuzz: #14583

Great, thanks!

> Have you tested a build that doesn't depend on the `icu4c` formula?

Even when removing the explicit `icu4c` dependency from the formula, it still gets downloaded as part of the installation process (presumably because `harfbuzz` [depends on it](https://github.com/Homebrew/homebrew-core/blob/b4f5f667591ed133cbf48a972ed2912c507b4de1/Formula/harfbuzz.rb#L30)).

Maybe @pkgw has some insights into the build failures and TeX errors.

**Comment by ilovezfs** [comment]
@iKevinY any updates here?

**Comment by iKevinY** [comment]
@ilovezfs I'll post on Tectonic's forum to see if anyone has suggestions on how to fix these issues.

**Comment by alexreg** [comment]
@ilovezfs Did you have any more luck with a minimal tex file?

**Comment by ilovezfs** [comment]
@alexreg Correct. It did work with a minimal tex file when had I tried that, just not for the paper.tex in the tectonic repo.

>the build fails on Yosemite, and also fails on El Capitan unless I have Xcode not just the CLT.

That is probably a higher priority problem than getting paper.tex to work :)

**Comment by pkgw** [comment]
Sorry, I'm really slammed at work right now and haven't been able to spend any time on this. Unfortunately it looks like this will continue to be the case for the next few weeks at least.

**Comment by ilovezfs** [comment]
@pkgw no worries! Thanks for the status update.

**Comment by alexreg** [comment]
Just a note: the *rust* dependency should be optional, to allow for e.g. *rustup* versions of Rust.

**Comment by pkgw** [comment]
Heads up: I've just published version 0.1.6 of the Tectonic rust crate.

**Comment by ilovezfs** [comment]
@pkgw neat! Note that https://github.com/tectonic-typesetting/tectonic/releases/latest still points to 0.1.5

**Comment by pkgw** [comment]
@ilovezfs Yeah — will update that after I write up a release announcement.

**Comment by ilovezfs** [comment]
@iKevinY would be cool to get the Homebrew formula into the release notes :) Wonder if we should just `depends_on :el_capitan` it for now.

**Comment by pkgw** [comment]
@iKevinY @ilovezfs There's a ton to write for the release notes and I need to revamp the website a bit too, so you probably have some time :-)

**Comment by ilovezfs** [comment]
@pkgw I'm still seeing this for `paper.tex`
```
iMac-TMP:paper joe$ tectonic paper.tex 
note: this is a BETA release; ask questions and report bugs at https://tectonic.newton.cx/
note: indexing https://purl.org/net/pkgwpub/tectonic-default
note: resolved to https://dl.bintray.com/pkgw/tectonic/tl2016extras/2016.0r4/tlextras-2016.0r4.tar
note: downloading SHA256SUM
note: generating format "latex"
Running TeX ...
error: something bad happened inside TeX; its output follows:

===============================================================================
(paper.tex
LaTeX2e <2016/03/31>
Babel <3.9r> and hyphenation patterns for 83 language(s) loaded.
(aastex6.cls

LaTeX Warning: You have requested document class `aastex6',
               but the document class provides `aastex'.

Document Class: aastex 2016/02/16 1.0/AAS markup document class
(times.sty) (lineno.sty) (revtex4-1.cls
Document Class: revtex4-1 2010/07/25/20:33:00 4.1r (http://publish.aps.org/revt
ex4/ for documentation)
ltxutil[2010/07/25/20:33:00 4.1r utilities package (portions licensed from W. E
. Baxter web at superscript.com)]
ltxfront[2010/07/25/20:33:00 4.1r frontmatter package (AO,DPC)]
ltxgrid[2010/07/25/20:33:00 4.1r page grid package (portions licensed from W. E
. Baxter web at superscript.com)]

Class revtex4-1 Warning: Failed to recognize \@vspace, \@vspacer, \@no@pgbk, \@
newline, and \\; no patches applied. Please get a more up-to-date class, .

(aps4-1.rtx) (aps10pt4-1.rtx) (textcase.sty) (url.sty) (natbib.sty)
(revsymb4-1.sty)) (latexsym.sty) (graphicx.sty (keyval.sty) (graphics.sty
(trig.sty) (graphics.cfg) (xetex.def (infwarerr.sty) (ltxcmds.sty))))
(amssymb.sty (amsfonts.sty)) (epsf.sty
This is `epsf.tex' v2.7.4 <14 February 2011>
) (longtable.sty) (hyperref.sty (hobsub-hyperref.sty (hobsub-generic.sty))
(ifxetex.sty) (auxhook.sty) (kvoptions.sty) (pd1enc.def) (hyperref.cfg))

Package hyperref Message: Driver (autodetected): hxetex.

(hxetex.def (puenc.def) (stringenc.sty) (rerunfilecheck.sty)) (xcolor.sty
(color.cfg)) (array.sty) (ulem.sty))===============================================================================

error: failed to open input file "mode"
```

**Comment by ilovezfs** [comment]
Is there a basic doc you like to test with it?

**Comment by pkgw** [comment]
@ilovezfs That's the one that I use. That's a strange failure, since the input file `mode.tex` should just be sitting right in the `tests/xenia` directory.

**Comment by ilovezfs** [comment]
Oh I'm just naively downloading the single file. Let me try cloning the full repo like a civilized person.

**Comment by ilovezfs** [comment]
Voilà 
[paper.pdf](https://github.com/Homebrew/homebrew-core/files/1133054/paper.pdf)

**Comment by pkgw** [comment]
Oh :-)

I think I'm seeing that there's a build failure on Yosemite relating to the identifier `z_const` not being defined. Presumably some zlib versioning problem and probably something that can be worked around in the source code pretty easily; something like:

```
#ifndef z_const
#define z_const const
#endif
```

Can you patch the source code in the Homebrew build system or anything like that?

**Comment by ilovezfs** [comment]
@pkgw yup. If you put something in a PR in https://github.com/tectonic-typesetting/tectonic, I can apply it as a patch here like this:
```
patch do
  url "https://github.com/tectonic-typesetting/tectonic/pull/1234.patch"
  sha256 "37e46ed0ee35a3f0957d847ce4afc871c352108279f8c001c7879282a8706495"
end
```

**Comment by pkgw** [comment]
@ilovezfs Well here's [a relevant Tectonic issue](https://github.com/tectonic-typesetting/tectonic/issues/102). Do @rekka's suggestions provide any hints about how to avoid the problem without requiring a patch?

**Comment by ilovezfs** [comment]
@pkgw I can try throwing `depends_on "zlib" if MacOS.version <= :yosemite` at it and see if the :spaghetti: sticks :)

**Comment by ilovezfs** [comment]
@pkgw yes that worked :)

**Comment by ilovezfs** [comment]
@pkgw so testing on OS X 10.11 El Capitan, without Xcode and only with the CLT, this is the failure that occurs: https://gist.github.com/ilovezfs/424e3449c1c324c1bb6c2829086c7966

Any ideas there?

**Comment by ilovezfs** [comment]
Hm looks like it's the same failure! Well that's convenient. That makes sense since the headers are different for zlib in the 10.11 CLT vs. Xcode 8.

**Comment by pkgw** [comment]
That looks like the same `z_const` problem, right? So I suspect that somehow that build configuration is causing the build system to find the older system zlib instead of the Homebrew version.

I saw a CI error because the `rubocop` tool was suspicious that the package depended on the `icu4c` formula. I haven't checked whether you can actually build Tectonic against the system version of ICU. Do you have a plan about how to deal with that problem? Presumably one needs to set some flag that says "yes I have checked that it's really necessary to use the Homebrew ICU".

**Comment by ilovezfs** [comment]
Xcode 8's headers are ZLIB_VERSION "1.2.8" whereas the CLT's ZLIB_VERSION is "1.2.5." It actually has nothing to do with Homebrew's zlib, which is a duplicate we avoid using when possible.

icu4c isn't an issue since we're depending on the icu4c formula, so the pkgconfig path for the icu4c will be automatically added to the build environment.

**Comment by ilovezfs** [comment]
@pkgw for simplicity I will just make it
```
depends_on "zlib" if MacOS.version <= :el_capitan
```

**Comment by ilovezfs** [comment]
@pkgw if you can manage at some point to drop the dependency for zlib from 1.2.8 to 1.2.5 that would be great, so that we can avoid the extra dependency, but that doesn't need to block the PR.

**Comment by ilovezfs** [comment]
>I saw a CI error because the rubocop tool was suspicious that the package depended on the icu4c formula. I haven't checked whether you can actually build Tectonic against the system version of ICU. Do you have a plan about how to deal with that problem? Presumably one needs to set some flag that says "yes I have checked that it's really necessary to use the Homebrew ICU".

@pkgw the check for use of Homebrew formulae that are duplicates of system libraries only run for `brew audit --new-formula` so it's not a problem once a maintainer has signed off on the initial formula, as it will not cause subsequent CI runs to fail when the formula is no longer new.

**Comment by pkgw** [comment]
Ah, OK. Zlib versioning issue filed as (tectonic #117)[https://github.com/tectonic-typesetting/tectonic/issues/117]. Other than that, we're almost there, right?

**Comment by ilovezfs** [comment]
@pkgw yeah, I'm pretty happy with it right now. Only a couple outstanding items.

1. I'm seeing a new failure, now that I've added the zlib dependency to get passed that, when building with just the CLT:

https://gist.github.com/ilovezfs/9a962fe8630a49b250f170942f265c52

2. Why is it specifying both `OPENSSL_INCLUDE_DIR` and `DEP_OPENSSL_INCLUDE`? Are they really both needed?

3. There was some skepticism (see https://github.com/tectonic-typesetting/tectonic/issues/4#issuecomment-308271563) about the desirability of using a test block that triggered a bunch of downloads, which the one I put in just now will do:

```
  test do
    system bin/"tectonic", "-o", testpath, pkgshare/"tests/xenia/paper.tex"
    assert_predicate testpath/"paper.pdf", :exist?, "Failed to create paper.pdf"
    assert_match "PDF document", shell_output("file paper.pdf")
  end
```
So we could either accept the downloads that come with that test, use a simpler document to limit them a bit more than paper.tex will, come up with an offline test that is non-trivial.

**Comment by alexreg** [comment]
Yes, I strongly support point 3 by @ilovezfs. Also, can we make the Rust dependency optional, please, for Rustup users?

**Comment by ilovezfs** [comment]
>Also, can we make the Rust dependency optional, please, for Rustup users?

Nope. It's needed at build time and there is no `depends_on :rust` requirement defined in brew at this point, only `depends_on "rust"` (the formula).

**Comment by alexreg** [comment]
> Also, can we make the Rust dependency optional, please, for Rustup users?
> Nope. It's needed at build time and there is no depends_on :rust requirement defined in brew at this point, only depends_on "rust" (the formula).

You misunderstand. If it's a `:recommended` dependency, then it allows users who use Rustup rust to bypass installing a Homebrew Rust, which is advantageous.

**Comment by alexreg** [comment]
BTW, when I try to install now, I get:
```
Error: /usr/local/opt/sqlite not present or broken
```

**Comment by pkgw** [comment]
1. Hmm, looks like some C++ standard library versioning issue. Maybe something like needing to link with `libc++` instead of `libstdc++`, that sort of thing? I'm not familiar with the details here but I bet some Googling will reveal people who have grappled with the same problem.

2. This is a bit of cargo-cult on my part but as far as I can tell it's needed. See [this lengthy GitHub issue](https://github.com/sfackler/rust-openssl/issues/255).

3. It does seem reasonable to avoid this. The Tectonic source tree includes all of the files needed to build the `plain` format, which is built and exercised in the standard test suite. If you're running the built-in tests with `cargo test` I don't think it's necessary to also run the `xenia` test.

**Comment by ilovezfs** [comment]
>You misunderstand. If it's a :recommended dependency, then it allows users who use Rustup rust to bypass installing a Homebrew Rust, which is advantageous.

rustup's rust won't be in the PATH. And it's not a concern because rust is a build-time only requirement anyway, so people who pour the pre-built bottle binary have nothing to worry about.

**Comment by pkgw** [comment]
@alexreg Strange! There is definitely nothing in Tectonic that has anything to do with `sqlite`, so it must be something on the Homebrew side.

**Comment by ilovezfs** [comment]
>Error: /usr/local/opt/sqlite not present or broken

This is a problem with your brew installation. Probably just need to
```
rm /usr/local/opt/sqlite
rm -rf /usr/local/Cellar/sqlite
brew install sqlite
```

**Comment by alexreg** [comment]
> Hmm, looks like some C++ standard library versioning issue. Maybe something like needing to link with libc++ instead of libstdc++, that sort of thing? I'm not familiar with the details here but I bet some Googling will reveal people who have grappled with the same problem.

Yes, looks like exactly that. C++11 vs. C++.

**Comment by alexreg** [comment]
> You misunderstand. If it's a :recommended dependency, then it allows users who use Rustup rust to bypass installing a Homebrew Rust, which is advantageous.
> rustup's rust won't be in the PATH. And it's not a concern because rust is a build-time only requirement anyway, so people who pour the pre-built bottle binary have nothing to worry about.

It's still extra disk-space, which is non-ideal. But I see your point about `PATH`. (Well, it *may* be in the `PATH`, but not necessarily.) Ideally we'd have something like `:rust`, but I know the Homebrew maintainer is strongly against these things, for some odd reason, which I forget.

**Comment by ilovezfs** [comment]
> If you're running the built-in tests with cargo test I don't think it's necessary to also run the xenia test.

@pkgw we cannot run `cargo test` in the test block because the build-time-only rust dependency will be unavailable at that point.

**Comment by pkgw** [comment]
@ilovezfs Ah, I see. If small downloads are acceptable, you could test with a nonsense document that uses the `plain` format, which requires many fewer resources. If you really want to avoid any downloads at all, that's trickier ... the distribution could perhaps embed a tiny bundle file with the resources needed to create the `plain` format.

I'm going to be offline for a while starting now. I still need to write the release notes so the announcement is very unlikely to go out before tomorrow at this rate.

**Comment by alexreg** [comment]
Just want to confirm: all builds fine for me now!

**Comment by ilovezfs** [comment]
>Just want to confirm: all builds fine for me now!

excellent!

**Comment by ilovezfs** [comment]
>I'm going to be offline for a while starting now. I still need to write the release notes so the announcement is very unlikely to go out before tomorrow at this rate.

hehe feel free to blame me :)

**Comment by ilovezfs** [comment]
OK. The build failure is fixed by adding
```
    ENV.cxx11
    ENV["MACOSX_DEPLOYMENT_TARGET"] = MacOS.version
```

**Comment by ilovezfs** [comment]
I am more than happy to review a PR for a lighter-weight, or even offline test, but I think this is good to ship now.

**Review by ilovezfs — APPROVED** [review]

**Comment by iKevinY** [comment]
Oh wow, what a nice surprise to see when checking my GitHub notifications! Thank you to everyone for getting this formula working! 😄

**Comment by ilovezfs** [comment]
Thanks for the :new: formula @iKevinY! Shipped!

Also, thanks @pkgw and @alexreg for the helping hands! It's much appreciated. :)

**Comment by alexreg** [comment]
Wonderful. Great work, guys!

**Comment by pkgw** [comment]
Great! I've [updated the website](https://tectonic-typesetting.github.io/en-US/install.html) to mention Homebrew as an option now. ([Relevant commit](https://github.com/tectonic-typesetting/tectonic-typesetting.github.io/commit/ddb1ca7d6337423956365c482dae57a9760058e9).)

**Comment by ilovezfs** [comment]
@pkgw awesome! Thanks :)

**Comment by ilovezfs** [comment]
@pkgw note that I made --with-graphviz the default for harfbuzz now for the sake of tectonic, so the source-build instructions on your site are slightly outdated now.

**Comment by ilovezfs** [comment]
s/graphviz/graphite2/

See https://github.com/Homebrew/homebrew-core/pull/14583

**Comment by pkgw** [comment]
Oh, nice, thanks! Can I trouble you to submit a PR on the website repo to make the language more accurate?

**Comment by ilovezfs** [comment]
@pkgw I'd be happy to.

**Comment by ilovezfs** [comment]
@pkgw https://github.com/tectonic-typesetting/tectonic-typesetting.github.io/pull/10

**Comment by pkgw** [comment]
@ilovezfs Re: [the zlib stuff](https://github.com/Homebrew/homebrew-core/pull/14575#issuecomment-313866038), [tectonic #120](https://github.com/tectonic-typesetting/tectonic/pull/120) may have fixed this on our end, so you may be able to drop the zlib dependency when the next version comes out.

**Comment by ilovezfs** [comment]
@pkgw yes it looks like it works to me:
```
iMac-TMP:~ joe$ mdutil -t `which tectonic`
/usr/local/Cellar/tectonic/HEAD-8201ce1/bin/tectonic
iMac-TMP:~ joe$ brew linkage tectonic
System libraries:
  /System/Library/Frameworks/AppKit.framework/Versions/C/AppKit
  /System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation
  /System/Library/Frameworks/CoreGraphics.framework/Versions/A/CoreGraphics
  /System/Library/Frameworks/CoreText.framework/Versions/A/CoreText
  /System/Library/Frameworks/Foundation.framework/Versions/C/Foundation
  /System/Library/Frameworks/Security.framework/Versions/A/Security
  /usr/lib/libSystem.B.dylib
  /usr/lib/libc++.1.dylib
  /usr/lib/libobjc.A.dylib
  /usr/lib/libz.1.dylib
Homebrew libraries:
  /usr/local/opt/freetype/lib/libfreetype.6.dylib (freetype)
  /usr/local/opt/graphite2/lib/libgraphite2.3.dylib (graphite2)
  /usr/local/opt/harfbuzz/lib/libharfbuzz-icu.0.dylib (harfbuzz)
  /usr/local/opt/harfbuzz/lib/libharfbuzz.0.dylib (harfbuzz)
  /usr/local/opt/icu4c/lib/libicudata.58.2.dylib (icu4c)
  /usr/local/opt/icu4c/lib/libicuuc.58.dylib (icu4c)
  /usr/local/opt/libpng/lib/libpng16.16.dylib (libpng)
iMac-TMP:~ joe$ brew config | grep Xcode
Xcode: N/A
iMac-TMP:~ joe$ 
```

**Pull request closed without merging**
